dali/python/nvidia/dali/shared_batch.py

```diff
@@ -0,0 +1,259 @@
+    ----------
+    func : callable
+        Function to be applied to every sample data object
+    sample : sample object or any nesting of those in tuple/list
+        Representation of sample
+    nest_with_sample: int
+        Specify how many consecutive (additional) arguments have the same level of nesting
+        as the sample.
+    """
+    if isinstance(sample, (tuple, list,)):
+        # Check that all the samples have common nesting
+        for i in range(nest_with_sample):
+            assert len(args[i]) == len(sample)
+        nest_group = sample, *args[0:nest_with_sample]
+        scalar_args = args[nest_with_sample:]
+        return type(sample)(_apply_to_sample(func, *part, *scalar_args) for part in zip(*nest_group))
+    else:
+        # we unpacked all nesting levels, now is actual data:
+        return func(sample, *args)
+
+
+class SharedBatchWriter:
+    """SharedBatchWriter can serialize and write batch into given shared
+    memory chunk (``mem_batch``).
+    """
+
+    SAMPLE_ALIGNMENT = 128
+    BUFFER_ALIGNMENT = 4096
+
+    def __init__(self, mem_batch: SharedMemChunk, batch):
+        import_numpy()
+        self.mem_batch = mem_batch
+        self.data_size = 0
+        self.meta_data_size = 0
+        self._write_batch(batch)
+
+    def _prepare_samples_meta(self, indexed_samples):
+        """Calculate metadata and total size of data to be serialized"""
+        data_size = 0
+
+        def make_meta(np_array):
+            nonlocal data_size
+            offset = _div_ceil(data_size, self.SAMPLE_ALIGNMENT) * self.SAMPLE_ALIGNMENT
+            data_size = offset + np_array.nbytes
+            return SampleMeta(offset, np_array.shape, np_array.dtype, np_array.nbytes)
+
+        meta = []
+        for idx, sample in indexed_samples:
+            meta.append((idx, _apply_to_sample(make_meta, sample)))
+        return meta, data_size
+
+    def _add_array_to_batch(self, np_array, meta, memview):
+        sample_size = meta.nbytes
+        offset = meta.offset
+        buffer = memview[offset:(offset + sample_size)]
+        shared_array = np.ndarray(
+            np_array.shape, dtype=np_array.dtype, buffer=buffer)
+        shared_array[:] = np_array[:]
+
+    def _write_batch(self, batch):
+        if not batch:
+            return
+        batch = [(idx, _apply_to_sample(_to_numpy, sample)) for idx, sample in batch]
+        meta, data_size = self._prepare_samples_meta(batch)
+        serialized_meta = pickle.dumps(meta)
+        self.meta_data_size = len(serialized_meta)
+        self.data_size = _div_ceil(data_size, self.SAMPLE_ALIGNMENT) * self.SAMPLE_ALIGNMENT
+        needed_capacity = self.data_size + self.meta_data_size
+        if self.mem_batch.capacity < needed_capacity:
+            grow_capacity = max(needed_capacity, 2 * self.mem_batch.capacity)
+            grow_capacity = _div_ceil(grow_capacity, self.BUFFER_ALIGNMENT) * self.BUFFER_ALIGNMENT
+            self.mem_batch.resize(grow_capacity)
```

Comment:
grow_capacity sounds like a boolean (should I grow?) or a delta (by how much should it grow?)
```suggestion
            new_capacity = max(needed_capacity, 2 * self.mem_batch.capacity)
            new_capacity = _div_ceil(new_capacity, self.BUFFER_ALIGNMENT) * self.BUFFER_ALIGNMENT
            self.mem_batch.resize(new_capacity)
```